### DISCUSSION TITLE
Feat: renew/refresh token

### DISCUSSION THREAD

**Discussion by jibon57** [8xFF/atm0s-media-server#384] [Ideas]
Token has a valid TTL so from application end need to renew it occasionally because during re-connection in various reason it will give error.

**Comment by giangndm** [comment]
Token is only used when:

- Connect to server
- Join to room

Then I think we can do it in application level, each time user want to connect to room, it can call to application server to get new token. Sample with join to room.

How do you think about above solution @jibon57

**Reply by jibon57** [reply]
What will be happen if ice need to restart: https://github.com/8xFF/atm0s-media-sdk-ts/blob/0d9fc1339efdd2495c4adfa51dfb619ff48d6516/packages/sdk-core/src/session.ts#L280? Yes, from application it can generate new/renew token. In that case need to expose one function in SDK to allow to update the token so that SDK can use that new token during restart/reconnect etc.

**Reply by giangndm** [reply]
Oh, I missed that case. Do you have any suggestions for it? Should we add some callback for getting new token on SDK side or simple overwrite the old one?

**Reply by jibon57** [reply]
Having something inside SDK will be better so that application end do not need to worry about it. From server side you may consider to create one API to renew token or can use Server event to receive renewed token internally. SDK will periodically send request either by http request or through data channel (more secure) for example: every 3 or 5 minutes to renew it. Server will check if token need renew then reply by Server event (more secure) & update local token.
